azure/services/scalesetvms/scalesetvms.go

```diff
@@ -73,211 +73,95 @@ func (s *Service) Reconcile(ctx context.Context) error {
 	ctx, log, done := tele.StartSpanWithLogger(ctx, "scalesetvms.Service.Reconcile")
 	defer done()
 
-	var (
-		resourceGroup = s.Scope.ResourceGroup()
-		vmssName      = s.Scope.ScaleSetName()
-		instanceID    = s.Scope.InstanceID()
-		providerID    = s.Scope.ProviderID()
-		isFlex        = s.Scope.OrchestrationMode() == infrav1.FlexibleOrchestrationMode
-	)
+	spec := s.Scope.ScaleSetVMSpec()
+	scaleSetVMSpec, ok := spec.(*ScaleSetVMSpec)
+	if !ok {
+		return errors.Errorf("%T is not of type ScaleSetVMSpec", spec)
+	}
 
+	reconciler := s.Reconciler
+	var getter azure.ResourceSpecGetter = scaleSetVMSpec
+	var result interface{}
+	var err error
 	// Fetch the latest instance or VM data. AzureMachinePoolReconciler handles model mutations.
-	if isFlex {
-		resourceID := strings.TrimPrefix(providerID, azureutil.ProviderIDPrefix)
-		log.V(4).Info("VMSS is flex", "vmssName", vmssName, "providerID", providerID, "resourceID", resourceID)
-		// Using VMSS Flex, so fetch by resource ID.
-		vm, err := s.VMClient.GetByID(ctx, resourceID)
+	if scaleSetVMSpec.IsFlex {
+		log.V(4).Info("VMSS is flex", "vmssName", scaleSetVMSpec.Name, "providerID", scaleSetVMSpec.ProviderID, "resourceID", scaleSetVMSpec.ResourceID)
+		getter, err = scaleSetVMSpecToVMSpec(*scaleSetVMSpec)
 		if err != nil {
-			if azure.ResourceNotFound(err) {
-				return azure.WithTransientError(errors.New("vm does not exist yet"), 30*time.Second)
-			}
-			return errors.Wrap(err, "failed getting vm")
+			return errors.Wrap(err, "failed to convert scaleSetVMSpec to vmSpec")
 		}
-		s.Scope.SetVMSSVM(converters.SDKVMToVMSSVM(vm, infrav1.FlexibleOrchestrationMode))
-		return nil
+		reconciler = s.VMReconciler
+	} else {
+		log.V(4).Info("VMSS is uniform", "vmssName", scaleSetVMSpec.Name, "providerID", scaleSetVMSpec.ProviderID, "instanceID", scaleSetVMSpec.InstanceID)
 	}
 
-	log.V(4).Info("VMSS is uniform", "vmssName", vmssName, "providerID", providerID, "instanceID", instanceID)
-	// Using VMSS Uniform, so fetch by instance ID.
-	instance, err := s.Client.Get(ctx, resourceGroup, vmssName, instanceID)
+	// Because getter.Parameters() returns nil, this will only return the existing resource if it exists.
```

Comment:
I couldn't quite understand this comment. Is there another way to explain it?

Reply:
In `CreateOrUpdateResource()` in async.go, we attempt to get the existing resource and get the return value of `Parameters()`. If `Parameters()` returns `nil`, we return the existing resource if it exists, and if it is not `nil` we attempt to create or update it. By forcing `Parameters()` to return `nil`, we're never going to create or update and just using it as a getter. See [this line](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/a0dcdea1b8168babca3fc8935345a8026f89d568/azure/services/async/async.go#L129).

Reply:
+1 on clarifying the comment here to make it clear that `CreateOrUpdateResource` never actually creates or updates the scale set vm